net-test/bin/faucet.sh

```diff
@@ -0,0 +1,294 @@
+
+MAX_BODY_LENGTH=65536
+FAUCET_AMOUNT="1.0"
+
+MODE="$1"
+BITCOIN_CONF="$2"
+
+set -uo pipefail
+
+exit_error() {
+   printf >&2 "$1"
+   exit 1
+}
+
+log() {
+   printf >&2 "%s\n" "$1"
+}
+
+http_ok() {
+   local CONTENT_LENGTH
+   local CONTENT_TYPE
+
+   CONTENT_LENGTH=$1
+   CONTENT_TYPE=$2
+   printf "HTTP/1.0 200 OK\r\nContent-Length: $CONTENT_LENGTH\r\nContent-Type: $CONTENT_TYPE\r\n\r\n"
+}
+
+http_401() {
+   printf "HTTP/1.0 401 Unsupported Method\r\n\r\n"
+}
+
+http_500() {
+   local ERR="$1"
+   local ERR_LEN=${#ERR}
+   printf "HTTP/1.0 500 Internal Server error\r\nContent-Length: $ERR_LEN\r\nContent-Type: text/plain\r\n\r\n$ERR"
+}
+
+http_404() {
+   printf "HTTP/1.0 404 Not Found\r\n\r\n"
+}
+
+get_ping() {
+   http_ok 5 "text/plain"
+   printf "alive"
+   return 0
+}
+
+get_bitcoin_ping() {
+   bitcoin-cli -conf="$BITCOIN_CONF" ping >/dev/null 2>&1
+   if [ $? -eq 0 ]; then 
+      local MSG="Bitcoind appears to be running"
+      http_ok ${#MSG} "text/plain"
+      echo "$MSG"
+      return 0
+   else
+      http_500 "Bitcoind appears to be stopped"
+      return 1
+   fi
+}
+
+get_balance() {
+   BALANCE="$(bitcoin-cli -conf="$BITCOIN_CONF" getbalance 2>&1)"
+   if [ $? -eq 0 ]; then
+      http_ok "${#BALANCE}" "text/plain"
+      echo "$BALANCE"
+      return 0
+   else
+      http_500 "$BALANCE"
+      return 1
+   fi
+}
+
+get_utxos() {
+   local ADDR="$1"
+   UTXOS="$(bitcoin-cli -conf="$BITCOIN_CONF" listunspent 1 1000000 "[\"$ADDR\"]" 2>&1)"
+   if [ $? -eq 0 ]; then 
+      http_ok ${#UTXOS} "application/json"
+      echo "$UTXOS"
+      return 0
+   else
+      http_500 "$UTXOS"
+      return 1
+   fi
+}
+
+get_confirmations() {
+   local TXID="$1"
+   CONFIRMATIONS="$(bitcoin-cli -conf="$BITCOIN_CONF" gettransaction "$TXID" | jq -r '.confirmations')"
+   RC=$?
+   if [ $RC -eq 0 ]; then 
+      http_ok ${#CONFIRMATIONS} "text/plain"
+      echo "$CONFIRMATIONS"
+      return 0
+   elif [ $RC -eq 1 ]; then 
+      http_500 "$CONFIRMATIONS"
+      return 1
+   else
+      http_404
+      return 2
+   fi
+}
+
+post_sendbtc() {
+   local ADDR
+
+   # format: address\n
+   read ADDR
+   TXID="$(bitcoin-cli -conf="$BITCOIN_CONF" sendtoaddress "$ADDR" "$FAUCET_AMOUNT" 2>&1)"
+   if [ $? -ne 0 ]; then
+      http_500 "$TXID"
+      return 1
+   fi
+
+   ERR="$(bitcoin-cli -conf="$BITCOIN_CONF" importaddress "$ADDR" 2>&1)"
+   if [ $? -ne 0 ]; then
+      http_500 "$ERR"
+      return 1
+   fi
+
+   http_ok ${#TXID} "text/plain"
+   echo "$TXID"
+   return 0
+}
+
+parse_request() {
+   local REQLINE
+   local VERB=""
+   local REQPATH=""
+   local CONTENT_TYPE=""
+   local CONTENT_LENGTH=0
+
+   while read REQLINE; do
+      # trim trailing whitespace
+      REQLINE="${REQLINE%"${REQLINE##*[![:space:]]}"}"
+      if [ -z "$REQLINE" ]; then
+         break
+      fi
+
+      # log "   reqline = '$REQLINE'"
+
+      TOK="$(echo "$REQLINE" | egrep "GET|POST" | sed -r 's/^(GET|POST)[ ]+([^ ]+)[ ]+HTTP\/1.(0|1)$/\1 \2/g')"
```

Comment:
the options used for `sed` (maybe not this line, but overall) are not supported on the mac os version.  Using `gsed` is a suitable replacement.

Reply:
Ah, I think `sed -r` is a GNU-specific flag.  Does your `sed` accept `sed -E`, which is (now) a POSIX-wide standard for the same thing?

Reply:
yes, -E for *modern* format is supported, but whether that's enough not sure, that's why I installed gsed to not have to worry